### ISSUE TITLE
storage: upsert rehydration can panic when the since is too advanced

### ISSUE THREAD

**Issue by aljoscha** [MaterializeInc/materialize#27345]
### What version of Materialize are you using?

version

### What is the issue?

This is a more specific version of https://github.com/MaterializeInc/materialize/issues/19097, where we now know in what part the panic is occuring.

Example nighly run that has the panic: https://buildkite.com/materialize/nightly/builds/7919

The basic mechanism of the bug is that the controller is releasing read handles but for distributed-systems reasons the cluster is still trying to render the source before it learns about it having been dropped.

We can use the approach introduced in https://github.com/MaterializeInc/materialize/pull/27323 to fix this as well.

**Comment by def-** [comment]
This is still occurring in 0dt parallel-workload test all the time: https://buildkite.com/materialize/nightly/builds/8771#0190e633-ec63-43b7-9ab0-fce441d13f6d
Full list: https://ci-failures.dev.materialize.com/?key=test-failures&tfFilters=%5B%7B%22id%22%3A%22build_date%22%2C%22value%22%3A%5Bnull%2Cnull%5D%7D%2C%7B%22id%22%3A%22issue%22%2C%22value%22%3A%22materialize%2F27345%22%7D%5D
@aljoscha  @teskje How concerned should we be? At the least I can't really enable the parallel-workload 0dt scenario until this is fixed since it seems to panic in each run: https://github.com/MaterializeInc/materialize/pull/28191

**Comment by def-** [comment]
^ This occurred in Parker's account, seemingly related to 0dt upgrades too.

**Comment by aljoscha** [comment]
@def- The sentry issue you posted above is about a `compute_import` not being valid because of the as_of, not an upsert source. Could be related but not strictly the same code path as upsert sources.

Side note, for context: this happend during the July 26 upgrade to staging, which is the first time we did a 0dt deployment, and we had quite a few problems there, with `envd` pods crash-looping. Could be that that caused read handles to be timed out, and thus clusters to fail when trying to render sources.

**Comment by teskje** [comment]
Yeah, that last one looks more like https://github.com/MaterializeInc/incidents-and-escalations/issues/39. We haven't made progress on this because we didn't know how to reproduce the panic. Looks like the parallel-workload test might be a reliable repro? Does that also work locally?

> Could be that that caused read handles to be timed out, and thus clusters to fail when trying to render sources.

If you are talking about the `compute_import` panic, I don't think that's it. The compute controller is keeping read holds for all its replicas, it doesn't rely on leases. Or... it shouldn't. Obviously there is some bug somewhere :/

**Comment by def-** [comment]
Reproduces locally like this:
```diff
diff --git a/misc/python/materialize/mzcompose/__init__.py b/misc/python/materialize/mzcompose/__init__.py
index 949dea9594..fe71793632 100644
--- a/misc/python/materialize/mzcompose/__init__.py
+++ b/misc/python/materialize/mzcompose/__init__.py
@@ -97,8 +97,8 @@ DEFAULT_SYSTEM_PARAMETERS = {
     "persist_stats_audit_percent": "100",
     "persist_txn_tables": "lazy",  # removed, but keep value for older versions
     "persist_use_critical_since_catalog": "true",
-    "persist_use_critical_since_snapshot": "false",  # Disabled because of 0dt upgrades, TODO: reenable
-    "persist_use_critical_since_source": "false",  # Disabled because of 0dt upgrades, TODO: reenable
+    "persist_use_critical_since_snapshot": "true",
+    "persist_use_critical_since_source": "true",
     "persist_part_decode_format": "row_with_validate",
     "statement_logging_default_sample_rate": "0.01",
     "statement_logging_max_sample_rate": "0.01",
```
`bin/mzcompose --find parallel-workload down && bin/mzcompose --find parallel-workload run default --scenario=0dt-deploy`

**Comment by aljoscha** [comment]
> Reproduces locally like this:
> 
> ```diff
> diff --git a/misc/python/materialize/mzcompose/__init__.py b/misc/python/materialize/mzcompose/__init__.py
> index 949dea9594..fe71793632 100644
> --- a/misc/python/materialize/mzcompose/__init__.py
> +++ b/misc/python/materialize/mzcompose/__init__.py
> @@ -97,8 +97,8 @@ DEFAULT_SYSTEM_PARAMETERS = {
>      "persist_stats_audit_percent": "100",
>      "persist_txn_tables": "lazy",  # removed, but keep value for older versions
>      "persist_use_critical_since_catalog": "true",
> -    "persist_use_critical_since_snapshot": "false",  # Disabled because of 0dt upgrades, TODO: reenable
> -    "persist_use_critical_since_source": "false",  # Disabled because of 0dt upgrades, TODO: reenable
> +    "persist_use_critical_since_snapshot": "true",
> +    "persist_use_critical_since_source": "true",
>      "persist_part_decode_format": "row_with_validate",
>      "statement_logging_default_sample_rate": "0.01",
>      "statement_logging_max_sample_rate": "0.01",
> ```
> 
> `bin/mzcompose --find parallel-workload down && bin/mzcompose --find parallel-workload run default --scenario=0dt-deploy`

@def- That one is expected, as soon as those `persist_use_critical_since_*` options are turned on anything that does 0dt deployments will run into that panic sooner or later. Mostly sooner... 😅 

But what I'm saying, and which I think @teskje confirmed, is that that last sentry link is a different issue, and one that also popped up in production. So it's likely not related to 0dt upgrades. And we haven't observed the original issue (related to upsert sources) since we changed CI to use the "correct" `persist_use_critical_since_foo` settings.

**Comment by aljoscha** [comment]
@def- Could we close this one? From above, it looks like the recent failure we observed was https://github.com/MaterializeInc/incidents-and-escalations/issues/39, which is related but different.

And the failures that happen when enabling the `persist_use_critical_since_*` flags are expected, and there's https://github.com/MaterializeInc/materialize/issues/28576 for tracking if/when/how we want to re-enable those.
